Bluetooth: tester: Classic: Enable bt tester for br l2cap

lylezhu2012 opened the pull request:
* Bluetooth: tester: Support BR L2CAP send data command

If the channel id is more than ARRAY_SIZE(channels), send L2CAP data through BR transport.

* Bluetooth: tester: Support L2CAP retransmission and flow-control mode

Configure the following configurations to support the L2CAP retransmission and flow-control mode.
CONFIG_BT_L2CAP_RET=y
CONFIG_BT_L2CAP_FC=y
CONFIG_BT_L2CAP_ENH_RET=y
CONFIG_BT_L2CAP_STREAM=y
CONFIG_BT_L2CAP_FCS=y
CONFIG_BT_L2CAP_EXT_WIN_SIZE=y
CONFIG_BT_L2CAP_MAX_WINDOW_SIZE=5

* Bluetooth: tester: Support BR L2CAP channel data receiving

Add the function `br_alloc_buf_cb()` to allocate receiving buffer for BR L2CAP channel.

Add the function `br_recv_cb()` to report the received BR L2CAP channel data.

* Bluetooth: tester: Support L2CAP BR ECHO req/rsp

Register ECHO callbacks in function `tester_init_l2cap()`.

Unregister ECHO callbacks in function `tester_unregister_l2cap()`.

Send ECHO response if ECHO request is received.

* Bluetooth: tester: Add L2CAP ECHO REQ BTP command

Add a BTP command `BTP_L2CAP_SEND_ECHO_REQ` to send L2CAP ECHO REQ.

* Bluetooth: tester: Add BTP command BTP_L2CAP_CONNECT_V2

Add a new BTP L2CAP command BTP_L2CAP_CONNECT_V2 to set extend the L2CAP connect request.
Compared with BTP command `connect`, two fields are extended, including `mode` and `options`.

* Bluetooth: tester: Add BTP command BTP_L2CAP_LISTEN_V2

Add a new BTP L2CAP command BTP_L2CAP_LISTEN_V2 to extend the L2CAP server feature.

Compared with BTP command `listen`, two fields are added, including `mode` and `options`.

* Bluetooth: tester: L2CAP: Set TX data buffer count to windows size

If `CONFIG_BT_CLASSIC` and `CONFIG_BT_L2CAP_MAX_WINDOW_SIZE` are defined, set the TX data buffer count to CONFIG_BT_L2CAP_MAX_WINDOW_SIZE. Or, set the TX data buffer count to CHANNELS.

* Bluetooth: tester: Classic: L2CAP: Support local busy

If the option BTP_L2CAP_CONNECT_V2_OPT_HOLD_CREDIT or BTP_L2CAP_LISTEN_OPT_HOLD_CREDIT is set, set the local flag
`hold_credit`. When the data received, holds the buffer and returns the error code `-EINPROGRESS`.

The held buffers will be released if the L2CAP channel disconnect callback is triggered. Or only of the held buffers will be released if the L2CAP BTP command `credits` is received.

* Bluetooth: tester: Classic: change L2CAP server count to 2

If CONFIG_BT_CLASSIC is enabled, change the count of L2CAP server to 2.

* Bluetooth: tester: Add command to send connectionless data

Set `CONFIG_BT_L2CAP_CLS` to enable L2CAP connectionless data reception feature.

Add command `BTP_L2CAP_CLS_SEND` to send connectionless data with specific PSM.

lylezhu2012 committed:
Bluetooth: tester: Support BR L2CAP send data command

If the channel id is more than ARRAY_SIZE(channels), send L2CAP data
through BR transport.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support L2CAP retransmission and flow-control mode

Configure the following configurations to support the L2CAP
retransmission and flow-control mode.
CONFIG_BT_L2CAP_RET=y
CONFIG_BT_L2CAP_FC=y
CONFIG_BT_L2CAP_ENH_RET=y
CONFIG_BT_L2CAP_STREAM=y
CONFIG_BT_L2CAP_FCS=y
CONFIG_BT_L2CAP_EXT_WIN_SIZE=y
CONFIG_BT_L2CAP_MAX_WINDOW_SIZE=5

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support BR L2CAP channel data receiving

Add the function `br_alloc_buf_cb()` to allocate receiving buffer for
BR L2CAP channel.

Add the function `br_recv_cb()` to report the received BR L2CAP
channel data.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support L2CAP BR ECHO req/rsp

Register ECHO callbacks in function `tester_init_l2cap()`.

Unregister ECHO callbacks in function `tester_unregister_l2cap()`.

Send ECHO response if ECHO request is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add L2CAP ECHO REQ BTP command

Add a BTP command `BTP_L2CAP_SEND_ECHO_REQ` to send L2CAP ECHO REQ.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: L2CAP: Set TX data buffer count to windows size

If `CONFIG_BT_CLASSIC` and `CONFIG_BT_L2CAP_MAX_WINDOW_SIZE` are
defined, set the TX data buffer count to
CONFIG_BT_L2CAP_MAX_WINDOW_SIZE. Or, set the TX data buffer count to
CHANNELS.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add BTP command BTP_L2CAP_CONNECT_V2

Add a new BTP L2CAP command BTP_L2CAP_CONNECT_V2 to set extend the
L2CAP connect request.
Compared with BTP command `connect`, two fields are extended,
including `mode` and `options`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Classic: change L2CAP server count to 2

If CONFIG_BT_CLASSIC is enabled, change the count of L2CAP server to 2.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add command to send connectionless data

Add command `BTP_L2CAP_CONNLESS_SEND` to send connectionless data with
specific PSM.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Classic: L2CAP: Support local busy

If the option BTP_L2CAP_CONNECT_V2_OPT_HOLD_CREDIT or
BTP_L2CAP_LISTEN_V2_OPT_HOLD_CREDIT is set, set the local flag
`hold_credit`. When the data received, holds the buffer and returns
the error code `-EINPROGRESS`.

The held buffers will be released if the L2CAP channel disconnect
callback is triggered. Or only of the held buffers will be released
if the L2CAP BTP command `credits` is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
@sjanc @jhedberg Who can help on this?

sjanc reviewed (commented):
overall looks OK to me,   please update btp definition with connect_v2 response and comment in code why only v1 is used (since both are same and this is handled in common code)

once bot completes run to double check we don't regress on LE I think this should be ready to go in

lylezhu2012 committed:
Bluetooth: tester: Add BTP command BTP_L2CAP_LISTEN_V2

Add a new BTP L2CAP command BTP_L2CAP_LISTEN_V2 to extend the L2CAP
server feature.

Compared with BTP command `listen`, two fields are added, including
`mode` and `options`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on tests/bluetooth/tester/src/btp/btp_l2cap.h:
```diff
@@ -89,6 +92,66 @@ struct btp_l2cap_disconnect_eatt_chans_cmd {
+	struct btp_l2cap_connect_cmd_common common;
+	uint8_t mode;
+	uint32_t options;
+} __packed;
```
Sure. Added comments.

lylezhu2012 replied:
removed structure `btp_l2cap_connect_cmd_common`

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/89178#issuecomment-3569222597, board: nrf54, estimated start time: 08:12:02, test case count: 99, estimated duration: 0:38:06<details><summary>Test cases to be run</summary>L2CAP/COS/CED/BI-05-C<br>L2CAP/COS/CED/BI-09-C<br>L2CAP/COS/CED/BI-11-C<br>L2CAP/COS/CED/BI-13-C<br>L2CAP/COS/CED/BI-16-C<br>L2CAP/COS/CED/BI-17-C<br>L2CAP/COS/CFC/BV-01-C<br>L2CAP/COS/CFC/BV-02-C<br>L2CAP/COS/CFC/BV-03-C<br>L2CAP/COS/CFC/BV-04-C<br>L2CAP/COS/CFC/BV-05-C<br>L2CAP/COS/ECFC/BV-01-C<br>L2CAP/COS/ECFC/BV-02-C<br>L2CAP/COS/ECFC/BV-03-C<br>L2CAP/COS/ECFC/BV-04-C<br>L2CAP/LE/CPU/BV-01-C<br>L2CAP/LE/CPU/BV-02-C<br>L2CAP/LE/CPU/BI-01-C<br>L2CAP/LE/CPU/BI-02-C<br>L2CAP/LE/REJ/BI-02-C<br>L2CAP/LE/CFC/BV-01-C<br>L2CAP/LE/CFC/BV-02-C<br>L2CAP/LE/CFC/BV-03-C<br>L2CAP/LE/CFC/BV-04-C<br>L2CAP/LE/CFC/BV-05-C<br>L2CAP/LE/CFC/BV-06-C<br>L2CAP/LE/CFC/BV-07-C<br>L2CAP/LE/CFC/BI-01-C<br>L2CAP/LE/CFC/BV-08-C<br>L2CAP/LE/CFC/BV-09-C<br>L2CAP/LE/CFC/BV-10-C<br>L2CAP/LE/CFC/BV-11-C<br>L2CAP/LE/CFC/BV-12-C<br>L2CAP/LE/CFC/BV-13-C<br>L2CAP/LE/CFC/BV-14-C<br>L2CAP/LE/CFC/BV-15-C<br>L2CAP/LE/CFC/BV-16-C<br>L2CAP/LE/CFC/BV-18-C<br>L2CAP/LE/CFC/BV-19-C<br>L2CAP/LE/CFC/BV-20-C<br>L2CAP/LE/CFC/BV-21-C<br>L2CAP/LE/CFC/BV-22-C<br>L2CAP/LE/CFC/BV-23-C<br>L2CAP/LE/CFC/BV-24-C<br>L2CAP/LE/CFC/BV-25-C<br>L2CAP/LE/CFC/BV-26-C<br>L2CAP/LE/CFC/BV-27-C<br>L2CAP/LE/CFC/BV-28-C<br>L2CAP/LE/CFC/BV-29-C<br>L2CAP/LE/CFC/BV-30-C<br>L2CAP/LE/CFC/BV-31-C<br>L2CAP/LE/CFC/BV-32-C<br>L2CAP/LE/CID/BI-01-C<br>L2CAP/ECFC/BV-01-C<br>L2CAP/ECFC/BV-02-C<br>L2CAP/ECFC/BV-03-C<br>L2CAP/ECFC/BV-04-C<br>L2CAP/ECFC/BV-06-C<br>L2CAP/ECFC/BV-07-C<br>L2CAP/ECFC/BI-01-C<br>L2CAP/ECFC/BI-02-C<br>L2CAP/ECFC/BV-08-C<br>L2CAP/ECFC/BV-09-C<br>L2CAP/ECFC/BV-10-C<br>L2CAP/ECFC/BV-11-C<br>L2CAP/ECFC/BV-12-C<br>L2CAP/ECFC/BV-13-C<br>L2CAP/ECFC/BV-14-C<br>L2CAP/ECFC/BV-15-C<br>L2CAP/ECFC/BV-16-C<br>L2CAP/ECFC/BV-17-C<br>L2CAP/ECFC/BV-18-C<br>L2CAP/ECFC/BV-19-C<br>L2CAP/ECFC/BV-20-C<br>L2CAP/ECFC/BV-21-C<br>L2CAP/ECFC/BV-22-C<br>L2CAP/ECFC/BV-23-C<br>L2CAP/ECFC/BI-03-C<br>L2CAP/ECFC/BV-24-C<br>L2CAP/ECFC/BV-25-C<br>L2CAP/ECFC/BI-04-C<br>L2CAP/ECFC/BV-26-C<br>L2CAP/ECFC/BV-27-C<br>L2CAP/ECFC/BV-28-C<br>L2CAP/ECFC/BV-29-C<br>L2CAP/ECFC/BV-30-C<br>L2CAP/ECFC/BV-31-C<br>L2CAP/ECFC/BV-32-C<br>L2CAP/ECFC/BV-33-C<br>L2CAP/ECFC/BV-34-C<br>L2CAP/ECFC/BV-35-C<br>L2CAP/ECFC/BV-38-C<br>L2CAP/ECFC/BV-39-C<br>L2CAP/ECFC/BV-41-C<br>L2CAP/ECFC/BI-05-C<br>L2CAP/ECFC/BI-06-C<br>L2CAP/ECFC/BI-07-C<br>L2CAP/ECFC/BV-80-C<br>L2CAP/TIM/BV-03-C</details>
